File: leetcode/weekly-contest/20240324/100245-每个字符最多出现两次的最长子字符串.py
from collections import Counter

class Solution:
    def maximumLengthSubstring(self, s: str) -> int:
        cnt = Counter()
        maxLen = -1
        j = 0
        for i, c in enumerate(s):
            while cnt[c] == 2:
                cnt[s[j]] -= 1
                j += 1
            cnt[c] += 1
            maxLen = max(maxLen, i - j + 1)

        return maxLen
            

File: problem-sets/leetcode-weekly-contest/20240324/100245-每个字符最多出现两次的最长子字符串.cpp
#include <string>
#include <unordered_map>
using std::string;
using std::unordered_map;

class Solution
{
public:
  int maximumLengthSubstring(string s)
  {
    unordered_map<char, int> map;
    int i, j;
    int len = 2;
    for (i = 0, j = 0; i < s.size(); ++i) {
      while (map[s[i]] == 2) {
        --map[s[j++]];
      }
      ++map[s[i]];
      if (i - j + 1 > len) {
        len = i - j + 1;
      }
    }
    return len;
  }
};